docs/plugins/index.md

```diff
@@ -0,0 +1,296 @@
+The engine  will endeavour to support previous versions of an API for as long as possible, but no guarentees will be made for compatibility across major versions of DOME.
+
+APIs are provided as a struct of function pointers, returned from:
+```
+void* DOME_getAPI(API_TYPE type, int API_VERSION)
+```
+
+## Core
+
+This API allows your plugin to register modules and provides some basic utilities.
+
+### Acquisition
+
+```
+DOME_API_v0* core = (DOME_API_v0*)DOME_getAPI(API_DOME, DOME_API_VERSION);
+```
+
+### Enums: 
+
+#### enum: DOME_Result
+
+Various methods return an enum of type `DOME_Result`, which indicates success or failure. These are the valid values:
+
+ * `DOME_RESULT_SUCCESS`
+ * `DOME_RESULT_FAILURE`
+ * `DOME_RESULT_UNKNOWN`
+
+
+### Methods
+
+#### method: registerModule
+```
+DOME_Result registerModule(DOME_Context ctx, const char* name, const char* moduleSource)`
+```
+This call registers module `name` with the source code `moduleSource`. You cannot register modules with the same name as DOME's internal modules. These are reserved.
+Returns `DOME_RESULT_SUCCESS` if the module was successfully registered, and `DOME_RESULT_FAILURE` otherwise.
+
+#### method: lockModule
+```
+void lockModule(DOME_Context ctx, const char* name)`
+```
+This marks the module `name` as locked, so that further functions cannot modify it. It is recommended to do this after you have registered all the methods for your module, however there is no requirement to.
+
+#### method: registerFn
+```
+DOME_Result registerFn(DOME_Context ctx, const char* name, const char* signature, DOME_ForeignFn method)`
+```
+Register `method` as the function to call for the foreign method specified by `signature` in the module `name`. Returns `DOME_RESULT_SUCCESS` if the function was successfully registered, and `DOME_RESULT_FAILURE` otherwise.
+
+The format for the `signature` string is as follows:
+ * `static` if the method is a static class method, followed by a space, otherwise both are omitted.
+ * `ClassName` for the class method being declared, followed by a period (`.`)
+ * `methodName` which is the name of the field/method being exposed.
+   - If this is a field getter, nothing else is needed.
+   - If this is a field setter, add `=(_)`
+   - If this is a method, then parenthesis and a comma seperated list of underscores (`_`) follow, for the number of arguments the method takes. Wren methods can only have up to 16 arguments.
+
+`DOME_ForeignFn` methods have the signature: `void method(WrenVM* vm)`
+
+#### method: registerBindFn
+```
+DOME_Result registerBindFn(DOME_Context ctx, const char* moduleName, DOME_BindClassFn fn)
+```
+Register a method to call when trying to resolve class allocators for the module `moduleName`. Returns `DOME_RESULT_SUCCESS` if the function was successfully registered, and `DOME_RESULT_FAILURE` otherwise.
+A `DOME_BindClassFn` has the signature: `WrenForeignClassMethods method(const char* className)`.
```

Comment:
What if a plugin registers multiple modules, wouldn't it be better if it could distinguish classes in them? I think that like the Wren API, we need to pass the module name, too.

Reply:
Each module has a separate bindFn, so the module name isn't necessary information here, unlike the WrenVM which has a single call back to resolve all modules.